internal/arduino/builder/internal/preprocessor/gcc.go

```diff
@@ -77,10 +78,42 @@ func GCC(
 	if err != nil {
 		return Result{}, err
 	}
-	stdout, stderr, err := proc.RunAndCaptureOutput(ctx)
 
-	// Append gcc arguments to stdout
-	stdout = append([]byte(fmt.Sprintln(strings.Join(args, " "))), stdout...)
+	stdout := bytes.NewBuffer(nil)
+	stderr := bytes.NewBuffer(nil)
 
-	return Result{args: proc.GetArgs(), stdout: stdout, stderr: stderr}, err
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	count := 0
+	stderrLimited := writerFunc(func(p []byte) (int, error) {
+		// Limit the size of the stderr buffer to 100KB
+		n, err := stderr.Write(p)
+		count += n
+		if count > 100*1024 {
+			cancel()
+			fmt.Fprintln(stderr, i18n.Tr("Compiler error output has been truncated."))
```

Comment:
```suggestion
			fmt.Fprintln(stderr, i18n.Tr("Compiler error output has been truncated."))
			cancel()
```
Probably I'd swap just in case the context cancellation unblocks the `WaitWithinContext` and returns the stderr bytes, before we appended the error string.